src/esthri/src/http_server.rs

```diff
@@ -346,6 +344,16 @@ async fn stream_object_to_archive<T: S3 + Send + Clone>(
     header.set_mode(ARCHIVE_ENTRY_MODE);
     header.set_mtime(obj_info.last_modified.timestamp() as u64);
     header.set_size(obj_info.size as u64);
+
+    // Appending the suffix to the path within the archive should
+    // hopefully make dealing with (transparently) compressed files less
+    // confusing
+    let path = if obj_info.is_esthri_compressed() {
+        format!("{}.gz", path)
+    } else {
+        path.to_string()
+    };
```

Comment:
> So the situation here is someone is requesting an archive that itself might contain transparently compressed files? Adding the extension makes sense otherwise yeah I could see people wondering why they can't open the file etc.

Yep, exactly!

> What you have here will actually allocated an owned value in both cases. The first one (adding the .gz) is unavoidable. For the second one though, the string is fine as it, but because the first branch returns a string you need to return in the second branch as well. An alternative would be something like

I did actually spend some time mulling over this bit of code as I didn't like having to do the allocation in `path.to_string()`. I was thinking of doing something like

```
    let tmp_string; 
    let path = if obj_info.is_esthri_compressed() {
        tmp_string = format!("{}.gz", path);
        &tmp_string
    } else {
        path
    };
```

Which I _think_ is roughly equivalent to the `Cow` approach (but probably uglier), but ultimately decided against it because I thought the performance penalty wasn't worth the uglification of the code :smiley: . 

The `Cow` pattern is cool though, thanks - seems like the right idiomatic (rustonic?) way of doing things. I'll update to use it. I have to say though, having the type named `Cow` is a bit confusing for my small brain when it's used for this purpose - if I saw `Cow`s in the code I'd immediately think that the code was trying to do clone on write optimization and then get confused when there's no mutation going on. I guess I'll know better from now on!

Reply:
> if I saw `Cow`s in the code I'd immediately think that the code was trying to do clone on write optimization and then get confused when there's no mutation going on

Yeah it's a little confusing because that is what `Cow` _is_ for, but it just so happens that it also solves a pretty prevalent problem in rust where you want to return some data from a scope (typically a function but in your case just a block) that is either owned or borrowed. This brief blog does a good job explaining the different usecases - https://deterministic.space/secret-life-of-cows.html